packages/hub/services/crank-nonce-lock.ts

```diff
@@ -0,0 +1,65 @@
+import { inject } from '@cardstack/di';
+import { Prisma } from '@prisma/client';
+import BN from 'bn.js';
+import config from 'config';
+import { addSeconds } from 'date-fns';
+import { Wallet } from 'ethers';
+import { nowUtc } from '../utils/dates';
+
+export default class CrankNonceLock {
+  prismaManager = inject('prisma-manager', { as: 'prismaManager' });
+  ethersProvider = inject('ethers-provider', { as: 'ethersProvider' });
+  CRANK_LOCK_KEY = 1;
+
+  private async getNonce(chainId: number, dbTransaction: Prisma.TransactionClient) {
+    let crankNonce = await dbTransaction.crankNonce.findUnique({
+      where: { chainId: chainId },
+    });
+    let one = new BN(1);
+    let nextNonce = crankNonce?.nonce ? new BN(crankNonce.nonce.toString()).add(one) : new BN(0);
+    let nonceTTL = Number(config.get('nonceTTL'));
+    let now = nowUtc();
+    if (nextNonce.cmp(one) <= 0 || (crankNonce && addSeconds(crankNonce.updatedAt, nonceTTL) <= now)) {
+      let crank = new Wallet(config.get('hubPrivateKey'));
+      let provider = this.ethersProvider.getInstance(chainId);
+      nextNonce = new BN(await provider.getTransactionCount(crank.address, 'latest'));
+    }
+
+    await dbTransaction.crankNonce.upsert({
+      where: { chainId: chainId },
+      create: {
+        chainId,
+        nonce: BigInt(nextNonce.toString()),
+      },
+      update: {
+        nonce: BigInt(nextNonce.toString()),
+      },
+    });
+
+    return nextNonce;
+  }
+
+  async withNonce(chainId: number, cb: (nonce: BN) => Promise<any>) {
+    let prisma = await this.prismaManager.getClient();
+    await prisma.$executeRaw`SELECT pg_advisory_lock(${this.CRANK_LOCK_KEY})`;
+    //Make database transaction here
+    //because if the execution failed the nonce should be rollback
+    try {
+      return await prisma.$transaction(
+        async (tx) => {
+          let nonce = await this.getNonce(chainId, tx);
+          return await cb(nonce);
+        },
+        { maxWait: 3000, timeout: 10000 }
+      );
+    } finally {
+      await prisma.$executeRaw`SELECT pg_advisory_unlock(${this.CRANK_LOCK_KEY})`;
```

Comment:
What will happen if the process exits before reaching this block, e.g. the app is restarted? Is the crank in a deadlock?

Reply:
I put the unlock process in the final block to ensure that the advisory lock will be unlocked no matter what happens in the execution process.

Reply:
The process can terminate during execution for many reasons (server crash, app crash) after claiming the lock and before reaching the `finally` block. Does that mean someone will have to manually unlock it on app restart?

Reply:
I added this line of code `await prisma.$executeRaw`SET lock_timeout TO '${timeout}s'`;` to have a timeout for locking, is this good enough to avoid deadlock?

Reply:
I think so!